pkg/server/handler.go

```diff
@@ -0,0 +1,367 @@
+	return func(w http.ResponseWriter, r *http.Request) {
+		http.ServeFile(w, r, "./webapp/templates/forbidden.html")
+	}
+}
+
+func (ctrl *Controller) decodeGoogleCallbackResponse(resp *http.Response) (string, error) {
+	type callbackResponse struct {
+		ID            string
+		Email         string
+		VerifiedEmail bool
+		Picture       string
+	}
+
+	var userProfile callbackResponse
+	err := json.NewDecoder(resp.Body).Decode(&userProfile)
+	if err != nil {
+		return "", err
+	}
+
+	return userProfile.Email, nil
+}
+
+func (ctrl *Controller) decodeGithubCallbackResponse(resp *http.Response) (string, error) {
+	type callbackResponse struct {
+		ID        int64
+		Email     string
+		Login     string
+		AvatarURL string
+	}
+
+	var userProfile callbackResponse
+	err := json.NewDecoder(resp.Body).Decode(&userProfile)
+	if err != nil {
+		return "", err
+	}
+
+	return userProfile.Login, nil
+}
+
+func (ctrl *Controller) decodeGitLabCallbackResponse(resp *http.Response) (string, error) {
+	type callbackResponse struct {
+		ID        int64
+		Email     string
+		Username  string
+		AvatarURL string
+	}
+
+	var userProfile callbackResponse
+	err := json.NewDecoder(resp.Body).Decode(&userProfile)
+	if err != nil {
+		return "", nil
+	}
+
+	return userProfile.Username, nil
+}
+
+func (ctrl *Controller) newJWTToken(name string) (string, error) {
+	claims := jwt.MapClaims{
+		"iat":  time.Now().Unix(),
+		"name": name,
+	}
+
+	if ctrl.config.LoginMaximumLifetimeDays > 0 {
+		claims["exp"] = time.Now().Add(time.Hour * 24 * time.Duration(ctrl.config.LoginMaximumLifetimeDays)).Unix()
+	}
+
+	jwtToken := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
+	tk, err := jwtToken.SignedString([]byte(ctrl.config.JWTSecret))
+	if err != nil {
+		return "", err
+	}
+
+	return tk, nil
+}
+
+func (ctrl *Controller) logAndRedirect(w http.ResponseWriter, r *http.Request, logString string, shouldInvalidateCookie bool) {
+	ctrl.log.Error(logString)
+	if shouldInvalidateCookie {
+		invalidateCookie(w, stateCookieName)
+	}
+
+	http.Redirect(w, r, "/forbidden", http.StatusTemporaryRedirect)
+	return
+}
+
+func (ctrl *Controller) callbackRedirectHandler(getAccountInfoURL string, oauthConf *oauth2.Config, decodeResponse decodeResponseFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		cookie, err := r.Cookie(stateCookieName)
+		if err != nil {
+			ctrl.logAndRedirect(w, r, "missing state cookie", false)
+			return
+		}
+
+		cookieState := cookie.Value
+		requestState := r.FormValue("state")
+
+		if requestState != cookieState {
+			ctrl.logAndRedirect(w, r, "invalid oauth state", false)
+			return
+		}
+
+		code := r.FormValue("code")
+		if code == "" {
+			ctrl.logAndRedirect(w, r, "code not found", true)
+			return
+		}
+
+		token, err := oauthConf.Exchange(r.Context(), code)
+		if err != nil {
+			ctrl.logAndRedirect(w, r, "Exchanging auth code for token failed with "+err.Error(), true)
```

Comment:
Well, we either can use `fmt.Sprintf`, or slightly modify the signature:
```
func (ctrl *Controller) logAndRedirect(w http.ResponseWriter, r *http.Request, format string, args ...interface{})
```
I think it's okay to invalidate state cookies on error regardless of anything (even if those are missing).

Reply:
In general only error can be passed so I'll remove this boolean parameter, pass error when it's available and log it with `WithError`